src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/AsyncMethodNamesShouldEndInAsync.cs

```diff
@@ -33,6 +36,47 @@ public abstract class AsyncMethodNamesShouldEndInAsyncAnalyzer : DiagnosticAnaly
 
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterSymbolAction(AnalyzeMethod, SymbolKind.Method);
+        }
+
+        private static void AnalyzeMethod(SymbolAnalysisContext context)
+        {
+            var method = (IMethodSymbol)context.Symbol;
+            if (NeedsAsyncSuffix(method, context.Compilation))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(Rule, method.Locations.First(), method.Name));
+            }
+        }
+
+        private static bool NeedsAsyncSuffix(IMethodSymbol method, Compilation compilation)
+        {
+            if (!method.ReturnType.IsAwaitableType() || method.Name.EndsWith("Async", StringComparison.Ordinal))
+            {
+                return false;
+            }
+
+            // If the method is overrides or implements a base method, renaming it will cause errors. We will rename
+            // such a method when and only when the virtual, abstract, or interface method it overrides is also renamed.
+            if (method.IsOverride || method.IsImplementationOfAnyInterfaceMember())
+            {
+                return false;
+            }
+
+            // If the method's definition is in another assembly, its name is beyond our control. Renaming references
+            // to it will not help.
+            var containingAssembly = method.OriginalDefinition.ContainingAssembly;
+            var currentAssembly = compilation.Assembly;
+            if (containingAssembly != currentAssembly)
+            {
+                return false;
+            }
+
+            // TODO: Do not fire the analyzer for unit test methods.
+            // TODO: Do not fire the analyzer when it could introduce a name conflict or overload resolution issues.
```

Comment:
I don't think we need to address either of these TODOs. If you still feel otherwise, please open a separate issue, and replace these TODO comments with the issue ID.